piano-dispatcher/src/main/java/com/batch/android/dispatcher/piano/PianoDispatcher.java

```diff
@@ -0,0 +1,463 @@
+    Event buildPianoOnSiteAdsEvent(Batch.EventDispatcher.Type type, Batch.EventDispatcher.Payload payload) {
+
+        String pianoOnSiteEventName;
+        if (isImpression(type)) {
+            pianoOnSiteEventName = EVENT_IMPRESSION;
+        } else if (isClick(type)) {
+            pianoOnSiteEventName = EVENT_CLICK;
+        } else {
+            return null;
+        }
+
+        HashMap<String, Object> params = new HashMap<String, Object>() {{
+            put(ON_SITE_TYPE, ON_SITE_TYPE_PUBLISHER);
+            put(ON_SITE_ADVERTISER, getSource(payload));
+            put(ON_SITE_CAMPAIGN, getCampaign(payload));
+            put(ON_SITE_FORMAT, getMedium(payload, type));
+        }};
+       return new Event(pianoOnSiteEventName, params);
+    }
+
+    /**
+     * Build a Piano Custom Event from a Batch Event
+     *
+     * @param type Batch event type
+     * @param payload Batch event payload
+     * @return The Piano event to send
+     */
+    @VisibleForTesting
+    Event buildPianoCustomEvent(@NonNull Batch.EventDispatcher.Type type,
+                                  @NonNull Batch.EventDispatcher.Payload payload) {
+
+        String eventName = getPianoEventName(type);
+        HashMap<String, Object> eventData = new HashMap<String, Object>() {{
+            put(CAMPAIGN, getCampaign(payload));
+            put(MEDIUM, getMedium(payload, type));
+            put(SOURCE, getSource(payload));
+            put(SOURCE_FORCE, true);
+        }};
+
+        String trackingId = payload.getTrackingId();
+        if (trackingId != null && !trackingId.isEmpty()) {
+            eventData.put(BATCH_TRACKING_ID, trackingId);
+        }
+
+        String content = getContent(payload);
+        if (content != null && !content.isEmpty()) {
+            eventData.put(CONTENT, content);
+        }
+
+        if (type.isMessagingEvent()) {
+            String webViewAnalyticsId = payload.getWebViewAnalyticsID();
+            if (webViewAnalyticsId != null && !webViewAnalyticsId.isEmpty()) {
+                eventData.put(BATCH_WEBVIEW_ANALYTICS_ID, webViewAnalyticsId);
+            }
+        }
+        return new Event(eventName, eventData);
+    }
+
+    /**
+     * Get the campaign label to send at Piano.
+     *
+     * First, check for an "at_campaign" tag in the custom payload, or in the deeplink.
+     * If not, check for an "utm_campaign" tag in the custom payload, or in the deeplink.
+     * If not, check if there is a TrackingID attached.
+     * If not, use {@link PianoDispatcher#BATCH_DEFAULT_CAMPAIGN}
+     *
+     * @param payload Batch event payload
+     * @return The campaign label.
+     */
+    @NonNull
+    @VisibleForTesting
+    String getCampaign(@NonNull Batch.EventDispatcher.Payload payload) {
+        String campaign = getTagFromPayload(payload, AT_CAMPAIGN);
+        if (campaign != null && !campaign.isEmpty()) {
+            return campaign;
+        }
+        campaign = getTagFromPayload(payload, UTM_CAMPAIGN);
```

Comment:
IMO le pontage des UTM doit etre facultatif (mais on par defaut)

Reply:
Pas sur de comprendre ce que tu veux dire ?

Reply:
L'utilisation des UTM dans Piano doit a mon avis etre parametrable

Reply:
C'est le cas sur leur sdk web: https://developers.atinternet-solutions.com/piano-analytics/data-collection/sdks/javascript#:~:text=Campaign-,enableUTMTracking,-UTM%20parameters%20collected

Reply:
Ouai ok, faudrais voir pour récupèrer cette info: https://developers.atinternet-solutions.com/piano-analytics/data-collection/how-to-send-events/marketing-campaigns#collect-utm-as-sources

Reply:
c'est une config locale au sdk web